sqlmesh/core/linter/definition.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import annotations
+
+
+from sqlmesh.core.config.linter import LinterConfig
+
+from sqlmesh.core.model import Model
+
+from sqlmesh.utils.errors import raise_config_error
+from sqlmesh.core.console import get_console
+from sqlmesh.core.linter.rule import RuleSet
+
+from sqlmesh.core.config.linter import RuleListType
+
+
+def gather_rules(all_rules: RuleSet, rule_names: RuleListType) -> RuleSet:
+    if rule_names == "ALL":
+        rule_names = list(all_rules.keys())
+
+    rules = set()
+    for rule_name in rule_names:
+        if rule_name not in all_rules:
+            raise_config_error(f"Rule {rule_name} could not be found")
+
+        rules.add(all_rules[rule_name])
+
+    return RuleSet(rules)
+
+
+class Linter:
+    def fill_from_config(self, all_rules: RuleSet, config: LinterConfig) -> Linter:
+        self.all_rules = all_rules
+
+        exclude_rules = gather_rules(self.all_rules, config.exclude_rules)
+        included_rules = self.all_rules.difference(exclude_rules)
+
+        self.rules = gather_rules(included_rules, config.rules)
+        self.warn_rules = gather_rules(included_rules, config.warn_rules)
+
+        if overlapping := self.rules.intersection(self.warn_rules):
+            overlapping_rules = ", ".join(rule for rule in overlapping.keys())
+            raise_config_error(f"Found overlapping rules [{overlapping_rules}] in lint config.")
```

Comment:
This error message isn't very clear, "overlapping" without additional context doesn't say much. We should say something along the lines of `f"Rules cannot simultaneously warn and raise an error: {overlapping_rules}."`